Do not generate enum-value checks for function call left-hand sides

tautschnig opened the pull request:
We already disabled these checks for assignment left-hand sides, but
need to do the same for function calls. Else we'd be asserting validity
before the value has been assigned.

Fixes: #6586

Co-authored-by: Ilia Levin <eli@literatecode.com>

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Support enum-typed constants in format_expr

We should not have to fall back to irept pretty printing for these.

tautschnig committed:
Do not generate enum-value checks for function call left-hand sides

We already disabled these checks for assignment left-hand sides, but
need to do the same for function calls. Else we'd be asserting validity
before the value has been assigned.

Fixes: #6586, #6451

Co-authored-by: Ilia Levin <eli@literatecode.com>

tautschnig committed:
__CPROVER_enum_is_in_range should not yield redundant assertions

The assertions generated by --enum-range-check must not apply to
the expansion of __CPROVER_enum_is_in_range, for those would assert
exactly the same. Also, those expressions are rather costly: for N enum
values, N assertions would be generated.

danielsn commented on regression/cbmc/enum_is_in_range/no_duplicate.c:
```diff
@@ -0,0 +1,19 @@
+  first,
+  second,
+  third,
+  fourth,
```
Consider testing non-contiguous  enums

tautschnig replied:
Done.

tautschnig committed:
Simplify disjunctions over equalities to intervals

Disjunctions over equalities of integer constants can be rewritten as
closed intervals. This will most commonly appear when expanding
enum_is_in_range.

This also surfaced a bug in to_integer, which used the wrong type
reference.
